dotnet-engine/UnleashEngine/UnleashEngine.cs

```diff
@@ -5,85 +5,95 @@
 
 namespace Unleash;
 
-public class Context
-{
-    public string? UserId { get; set; }
-    public string? SessionId { get; set; }
-    public string? RemoteAddress { get; set; }
-    public string? Environment { get; set; }
-    public string? AppName { get; set; }
-    public string? CurrentTime { get; set; }
-    public Dictionary<string, string>? Properties { get; set; }
-}
-
-public class Variant
-{
-    public string Name { get; set; }
-    public Payload? Payload { get; set; }
-    public bool Enabled { get; set; }
-}
-
-public class Payload
-{
-    public string? PayloadType { get; set; }
-    public string? Value { get; set; }
-}
-
-
 public class UnleashEngine
 {
     private JsonSerializerOptions options = new JsonSerializerOptions
     {
         PropertyNamingPolicy = JsonNamingPolicy.CamelCase
     };
 
-    private IntPtr state;
-    private const string DLL_PATH = "../../../../../target/release/libyggdrasilffi.so";
-
-    [DllImport(DLL_PATH)]
-    private static extern IntPtr engine_new();
-
-    [DllImport(DLL_PATH)]
-    private static extern void engine_free(IntPtr ptr);
-
-    [DllImport(DLL_PATH)]
-    private static extern IntPtr engine_take_state(IntPtr ptr, string json);
-
-    [DllImport(DLL_PATH)]
-    private static extern bool engine_is_enabled(IntPtr ptr, string toggle_name, string context);
-
-    [DllImport(DLL_PATH)]
-    private static extern IntPtr engine_get_variant(IntPtr ptr, string toggle_name, string context);
+    private IFFIAccess platformEngine;
+
+    private static IFFIAccess GetPlatformEngine() { 
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) {
+            return new FFILinux();
+        } else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) {
+            return new FFIMacOS();
+        } else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
+            return new FFIWin();
+        } else {
+            throw new PlatformNotSupportedException();
+        }
+    }
 
-    [DllImport(DLL_PATH)]
-    private static extern void engine_free_variant_def(IntPtr ptr);
+    private IntPtr state;
 
     public UnleashEngine()
     {
-        this.state = engine_new();
+        platformEngine = GetPlatformEngine();
+        state = platformEngine.NewEngine();
     }
 
     public void Dispose()
     {
-        engine_free(this.state);
+        platformEngine.FreeEngine(this.state);
         GC.SuppressFinalize(this);
     }
 
     public void TakeState(string json)
     {
-        engine_take_state(state, json);
+        var takeStatePtr = platformEngine.TakeState(state, json);
+
+        if (takeStatePtr == IntPtr.Zero)
+        {
+            return;
+        }
+
+        var takeStateJson = Marshal.PtrToStringUTF8(takeStatePtr);
+
+        var takeStateResult = takeStateJson != null ?
+            JsonSerializer.Deserialize<EngineResponse>(takeStateJson, options) :
+            null;
+
+        platformEngine.FreeResponse(takeStatePtr);
```

Comment:
These free responses need to be as close to the reading of the pointer as possible 

(looks to be this line: `var takeStateJson = Marshal.PtrToStringUTF8(takeStatePtr);`)

Line 54 can potentially fail, which will cause a leak, as written, can we move this line to line 53? If we can't, we probably gotta wrap this in a try/finally

Same thing for other methods that release the pointers

Reply:
Thank you, valuable input! I'm fairly certain it can be moved!